File: 2024/python/day3.py
import unittest
import re


def part1(path):
    res = 0
    with open(path) as file:
        for match in re.findall(r"mul\(([0-9]+),([0-9]+)\)", file.read()):
            res += int(match[0]) * int(match[1])
    return res


def part2(path):
    res = 0
    enabled = True
    with open(path) as file:
        for match in re.findall(
            r"mul\(([0-9]+),([0-9]+)\)|(do)\(\)|(don't)\(\)", file.read()
        ):
            if match[2]:
                enabled = True
            elif match[3]:
                enabled = False
            elif enabled:
                res += int(match[0]) * int(match[1])
    return res


class TestDay1(unittest.TestCase):
    def test_part1(self):
        self.assertEqual(part1("../data/sample/day3part1.txt"), 161)

    def test_part2(self):
        self.assertEqual(part2("../data/sample/day3part2.txt"), 48)


if __name__ == "__main__":
    print("part 1:", part1("../data/day3.txt"))
    print("part 2:", part2("../data/day3.txt"))

    unittest.main()

File: 2024/zig/day3.zig
// run with -Ilib to include regex_wrapper.h

const std = @import("std");
const print = std.debug.print;
const re = @cImport(@cInclude("regex_wrapper.h"));

const input = @embedFile("data/day3.txt");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = gpa.allocator();

    const slice = try alloc.alignedAlloc(u8, re.alignof_regex_t, re.sizeof_regex_t);
    defer alloc.free(slice);
    const regex: [*]re.regex_t = @ptrCast(slice.ptr);

    if (re.regcomp(regex, "mul\\(([0-9]+),([0-9]+)\\)|(do)\\(\\)|(don't)\\(\\)", re.REG_EXTENDED | re.REG_ICASE) != 0) {
        return;
    }

    var matches: [5]re.regmatch_t = undefined;
    var offset: u64 = 0;
    var part1: u64 = 0;
    var part2: u64 = 0;
    var enabled: bool = true;
    while (re.regexec(regex, input[offset..], matches.len, &matches, 0) == 0) {
        if (matches[3].rm_so >= 0) {
            enabled = true;
        } else if (matches[4].rm_so >= 0) {
            enabled = false;
        }
        const a = try match_to_int(matches, 1, offset);
        const b = try match_to_int(matches, 2, offset);
        offset += @intCast(matches[0].rm_eo);
        part1 += a * b;
        if (enabled) {
            part2 += a * b;
        }
    }
    print("Part 1: {d}\n", .{part1});
    print("Part 2: {d}\n", .{part2});
}

fn match_to_int(matches: [5]re.regmatch_t, i: usize, offset: u64) !u64 {
    const m = matches[i];
    if (m.rm_so < 0) {
        return 0;
    }
    const rm_so: u64 = @intCast(m.rm_so);
    const rm_eo: u64 = @intCast(m.rm_eo);
    const start = rm_so + offset;
    const end = rm_eo + offset;
    if (start >= input.len or end >= input.len) {
        return 0;
    }
    const match = input[start..end];
    const num = try std.fmt.parseInt(u64, match, 10);
    return num;
}
